### ISSUE TITLE
[feature request]Implement of `Circle Loss`

### ISSUE THREAD

**Issue by DevinXYang** [THUFutureLab/gluon-face#38]
A new loss function for representation learning. [Circle Loss: A Unified Perspective of Pair Similarity Optimization](https://arxiv.org/pdf/2002.10857.pdf)
Just remind me to write.
And I may fix `FP16` training bugs if anyone could success training with it, or I'll drop this. I'm not sure mxnet support mixed precision training.

**Comment by wms2537** [comment]
I haven't tried circle loss yet. However, for the Arcloss to work in float16, you need to lower the feature scale s to 16. I think it is because when you use float16, the embedding space will change.

**Comment by wms2537** [comment]
However, I did a trick which is adding focal loss to scale the loss. Then, I could put the scale s to a larger value but that yields unstable training.

**Comment by DevinXYang** [comment]
Do you use `batch_size` of 512 to train? Smaller `batch_size` will lead to unconverge. I think you may notice this
```python
logger.info("Warning: If you train a model on emore with batch size < 512 may lead to not converge.You may try a smaller dataset.")
```
Or another reason may caused by `Mxnet` doesn't support mixed precision training.
I used to successfully train `ArcLoss` with large s , though not use `emore` dataset.

**Comment by DevinXYang** [comment]
And please let me know what part of the code you change to train with FP16, you could make a pr or just show here.

**Comment by wms2537** [comment]
I've used a batch size of 512*4, or 1024*4 for small networks such as resnet-18 as I have 4 GPUs. Below are the part I've changed for it to train successfully at fp16.
`margin_s = 16`

At this point the training should work fine.
For the loss scaling part:
`class ArcLoss(gluon.loss.Loss):
    r"""ArcLoss from
    `"ArcFace: Additive Angular Margin Loss for Deep Face Recognition"
    <https://arxiv.org/abs/1801.07698>`_ paper.
    Parameters
    ----------
    classes: int.
        Number of classes.
    m: float.
        Margin parameter for loss.
    s: int.
        Scale parameter for loss.
    Outputs:
        - **loss**: loss tensor with shape (batch_size,). Dimensions other than
          batch_axis are averaged out.
    """

    def __init__(self, classes, m=0.5, s=64, easy_margin=True, dtype="float32", sparse_label=True, weight=None, batch_axis=0, **kwargs):
        super().__init__(weight=weight, batch_axis=batch_axis, **kwargs)
        assert s > 0.
        assert 0 <= m < (math.pi / 2)
        self.s = s
        self.m = m
        self.cos_m = math.cos(m)
        self.sin_m = math.sin(m)
        self.mm = math.sin(math.pi - m) * m
        self.threshold = math.cos(math.pi - m)
        self._classes = classes
        self.easy_margin = easy_margin 
        self._dtype = dtype
        self._sparse_label = sparse_label
        self._axis = -1
        self._gamma = 2
        self._alpha = 0.25
        self._weight = weight

    def hybrid_forward(self, F, pred, label, sample_weight=None, *args, **kwargs):
        #print(pred, label)
        cos_t = F.pick(pred, F.argmax(label, axis=-1), axis=1)  # cos(theta_yi)
        if self.easy_margin:
            cond = F.Activation(data=cos_t, act_type='relu')
        else:
            cond_v = cos_t - self.threshold
            cond = F.Activation(data=cond_v, act_type='relu')

        # sin_t = F.sqrt(1.0 - cos_t * cos_t)  # sin(theta)
        # new_zy = cos_t * self.cos_m - sin_t * self.sin_m  # cos(theta_yi + m)

        new_zy = F.cos(F.arccos(cos_t) + self.m)  # cos(theta_yi + m)
        if self.easy_margin:
            zy_keep = cos_t
        else:
            zy_keep = cos_t - self.mm  # (cos(theta_yi) - sin(pi - m)*m)
        new_zy = F.where(cond, new_zy, zy_keep)
        diff = new_zy - cos_t  # cos(theta_yi + m) - cos(theta_yi)

        diff = F.expand_dims(diff, 1)  # shape=(b, 1)
        #gt_one_hot = F.one_hot(label, depth=self._classes, on_value=1.0, off_value=0.0, dtype=self._dtype)
        body = F.broadcast_mul(label, diff)
        pred = pred + body
        pred = pred * self.s

        
        log_softmax = F.log_softmax
        pick = F.pick
        pred = log_softmax(pred)
        pred = self._alpha*pred*(1-F.exp(pred))**self._gamma
        #print(pred, pred2)
        if self._sparse_label:
            loss = -pick(pred, label, axis=self._axis, keepdims=True)
        else:
            label = _reshape_like(F, label, pred)
            loss = -(pred * label).sum(axis=self._axis, keepdims=True)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        if mx.util.is_np_array():
            if F is ndarray:
                return loss.mean(axis=tuple(range(1, loss.ndim)))
            else:
                return F.npx.batch_flatten(loss).mean(axis=1)
        else:
            return loss.mean(axis=self._batch_axis, exclude=True)`

My code may not work immediately for you as I did the one-hot encoding out of the loss function separately.

**Comment by DevinXYang** [comment]
ok, thx.

**Comment by wms2537** [comment]
I've used a batch size of 512x4, or 1024x4 for small networks such as resnet-18 as I have 4 GPUs. Below are the part I've changed for it to train successfully at fp16.
margin_s = 16

At this point the training should work fine.
For the loss scaling part:
`class ArcLoss(gluon.loss.Loss):
    r"""ArcLoss from
    "ArcFace: Additive Angular Margin Loss for Deep Face Recognition"
    <https://arxiv.org/abs/1801.07698>_ paper.
    Parameters
    ----------
    classes: int.
        Number of classes.
    m: float.
        Margin parameter for loss.
    s: int.
        Scale parameter for loss.
    Outputs:
        - **loss**: loss tensor with shape (batch_size,). Dimensions other than
          batch_axis are averaged out.
    """

    def __init__(self, classes, m=0.5, s=64, easy_margin=True, dtype="float32", sparse_label=True, weight=None, batch_axis=0, **kwargs):
        super().__init__(weight=weight, batch_axis=batch_axis, **kwargs)
        assert s > 0.
        assert 0 <= m < (math.pi / 2)
        self.s = s
        self.m = m
        self.cos_m = math.cos(m)
        self.sin_m = math.sin(m)
        self.mm = math.sin(math.pi - m) * m
        self.threshold = math.cos(math.pi - m)
        self._classes = classes
        self.easy_margin = easy_margin 
        self._dtype = dtype
        self._sparse_label = sparse_label
        self._axis = -1
        self._gamma = 2
        self._alpha = 0.25
        self._weight = weight

    def hybrid_forward(self, F, pred, label, sample_weight=None, *args, **kwargs):
        #print(pred, label)
        cos_t = F.pick(pred, F.argmax(label, axis=-1), axis=1)  # cos(theta_yi)
        if self.easy_margin:
            cond = F.Activation(data=cos_t, act_type='relu')
        else:
            cond_v = cos_t - self.threshold
            cond = F.Activation(data=cond_v, act_type='relu')

        # sin_t = F.sqrt(1.0 - cos_t * cos_t)  # sin(theta)
        # new_zy = cos_t * self.cos_m - sin_t * self.sin_m  # cos(theta_yi + m)

        new_zy = F.cos(F.arccos(cos_t) + self.m)  # cos(theta_yi + m)
        if self.easy_margin:
            zy_keep = cos_t
        else:
            zy_keep = cos_t - self.mm  # (cos(theta_yi) - sin(pi - m)*m)
        new_zy = F.where(cond, new_zy, zy_keep)
        diff = new_zy - cos_t  # cos(theta_yi + m) - cos(theta_yi)

        diff = F.expand_dims(diff, 1)  # shape=(b, 1)
        #gt_one_hot = F.one_hot(label, depth=self._classes, on_value=1.0, off_value=0.0, dtype=self._dtype)
        body = F.broadcast_mul(label, diff)
        pred = pred + body
        pred = pred * self.s

        
        log_softmax = F.log_softmax
        pick = F.pick
        #if not self._from_logits:
        #pred = F.softmax(pred)
        #pred = F.log(pred)
        pred = log_softmax(pred)
        pred = self._alpha*pred*(1-F.exp(pred))**self._gamma
        #print(pred, pred2)
        if self._sparse_label:
            loss = -pick(pred, label, axis=self._axis, keepdims=True)
        else:
            label = _reshape_like(F, label, pred)
            loss = -(pred * label).sum(axis=self._axis, keepdims=True)
        loss = _apply_weighting(F, loss, self._weight, sample_weight)
        
        return loss.mean(axis=self._batch_axis, exclude=True)`
My code may not work immediately for you as I did the one-hot encoding out of the loss function separately.

**Comment by wms2537** [comment]
By the way, just for your information, I trained the model directly from the start to end with Arcloss, It converged as usual without a problem.

**Comment by DevinXYang** [comment]
You mean use a small `batch_size`?

**Comment by wms2537** [comment]
I haven't tried a small `batch_size` before as it takes a lot longer train. The `batch_size` I used is 512 per GPU card.

**Comment by DevinXYang** [comment]
OK, Thanks for your help.

**Comment by DevinXYang** [comment]
I've finished `Pytorch` version.
When I have devices to train this I'll implement it by `Gluon`.
